pkg/kubectl/cmd/set/set_selector.go

```diff
@@ -0,0 +1,280 @@
+		kubectl set selector deployment --all 'environment in (production, qa)'
+
+		# Update a rs identified by the type and name in "rs.json"
+		kubectl set selector -f rs.json 'environment notin (frontend)'
+
+		# Update service 'foo' only if the resource is unchanged from version 1.
+		kubectl set selector service foo status=unhealthy --resource-version=1`)
+)
+
+// NewCmdSelector is the "set selector" command.
+func NewCmdSelector(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &SelectorOptions{
+		Out: out,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "selector [--overwrite] (-f FILENAME | TYPE NAME) EXPRESSIONS [--resource-version=version]",
+		Short:   "Update the selectors on a resource",
+		Long:    fmt.Sprintf(selectorLong),
+		Example: selectorExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.Complete(f, cmd, args))
+			cmdutil.CheckErr(options.Validate())
+			cmdutil.CheckErr(options.Run())
+		},
+	}
+	cmdutil.AddPrinterFlags(cmd)
+	cmd.Flags().Bool("all", false, "Select all resources in the namespace of the specified resource types")
+	cmd.Flags().String("resource-version", "", "If non-empty, the selectors update will only succeed if this is the current resource-version for the object. Only valid when specifying a single resource.")
+	usage := "Filename, directory, or URL to a file identifying the resource to update the selectors"
+	cmd.Flags().Bool("overwrite", false, "If true, allow selectors to be overwritten, otherwise reject selector updates that overwrite existing selectors.")
+	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
+	cmdutil.AddRecursiveFlag(cmd, &options.Recursive)
+	cmdutil.AddDryRunFlag(cmd)
+	cmdutil.AddRecordFlag(cmd)
+	cmdutil.AddInclude3rdPartyFlags(cmd)
+
+	return cmd
+}
+
+// Complete assigns the SelectorOptions from args.
+func (o *SelectorOptions) Complete(f *cmdutil.Factory, cmd *cobra.Command, args []string) error {
+	o.Mapper, o.Typer = f.Object(cmdutil.GetIncludeThirdPartyAPIs(cmd))
+	o.UpdateSelectorForObject = f.UpdateSelectorForObject
+	o.Encoder = f.JSONEncoder()
+	o.ShortOutput = cmdutil.GetFlagString(cmd, "output") == "name"
+	o.Record = cmdutil.GetRecordFlag(cmd)
+	o.DryRun = cmdutil.GetDryRunFlag(cmd)
+	o.Overwrite = cmdutil.GetFlagBool(cmd, "overwrite")
+	o.ChangeCause = f.Command()
+	o.PrintObject = f.PrintObject
+	o.Cmd = cmd
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	o.Resources, o.Selector, err = getResourcesAndSelector(args)
+	if err != nil {
+		return err
+	}
+
+	builder := resource.NewBuilder(o.Mapper, o.Typer, resource.ClientMapperFunc(f.ClientForMapping), f.Decoder(true)).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, o.Recursive, o.Filenames...).
+		Flatten()
+	if !o.Local {
+		builder = builder.
+			ResourceTypeOrNameArgs(o.All, o.Resources...).
+			Latest()
+	}
+	o.Infos, err = builder.Do().Infos()
```

Comment:
This might make remote calls.  Try to refactor so the remote lookup happens in `Run`